text/0000-ques-in-main.md

```diff
@@ -0,0 +1,514 @@
+unchanged, and the shim should simply ignore whichever aspects of
+`Termination` don't make sense in context.
+
+There are also environments where
+[returning from `main` constitutes a _bug_.][divergent-main] If you
+are implementing an operating system kernel, for instance, there may
+be nothing to return to.  Then you want it to be a compile-time error
+for `main` to return anything other than `!`.  If everything is
+implemented correctly, such environments should be able to get that
+effect by omitting all stock impls of `Termination` other than for
+`!`.  Perhaps there should also be a compiler hook that allows such
+environments to refuse to let you impl Termination yourself.
+
+[divergent-main]: https://internals.rust-lang.org/t/allowing-for-main-to-be-divergent-in-embedded-environments/4717
+
+## Test functions and doctests
+[test-functions-and-doctests]: #test-functions-and-doctests
+
+The harness for `#[test]` functions is very simple; I think it would
+be enough to just give `#[test]` functions the same shim that we give
+to `main`.  The programmer would be responsible for adjusting their
+`#[test]` functions' return types if they want to use `?`, but
+existing code would continue to work.
+
+Doctests require a little magic, because you normally don't write the
+function head for a doctest yourself, only its body.  This magic
+belongs in rustdoc, not in rustc.  When `maketest` sees that it needs
+to insert a function head for `main`, it should now write out
+
+``` rust
+fn main () -> Result<(), ErrorT> {
+   ...
+   Ok(())
+}
+```
+
+for some value of `ErrorT` TBD.  It doesn't need to parse the body of
+the test to know whether it should do this; it can just do it
+unconditionally.
+
+## Standard impls of Termination
+[standard-impls-of-termination]: #standard-impls-of-termination
+
+At least the following implementations of Termination are available in
+libstd.  I use the ISO C constants `EXIT_SUCCESS` and `EXIT_FAILURE`
+for exposition.  [See below][unix-specific-refinements] for more
+discussion of these constants and the `TermStatus` type.
+
+``` rust
+impl Termination for ! {
+    fn write_diagnostic(&self, progname: &str, stream: &mut Write) -> ()
+    { unreachable!(); }
+    fn exit_status(&self) -> i32
+    { unreachable!(); }
+}
+
+impl Termination for () {
+    fn write_diagnostic(&self, progname: &str, stream: &mut Write) -> ()
+    { }
+    fn exit_status(&self) -> i32
+    { EXIT_SUCCESS }
+}
+
+impl Termination for TermStatus {
+    fn write_diagnostic(&self, progname: &str, stream: &mut Write) -> ()
+    { }
+    fn exit_status(&self) -> i32
+    { self.0 }
+}
+
+impl<E: Error> Termination for E {
+    fn write_diagnostic(&self, progname: &str, stream: &mut Write) -> ()
+    {
+        // unspecified, but not entirely unlike this:
+        if let Some(ref cause) = self.cause() {
+            cause.write_diagnostic(progname, stream);
+        }
+        writeln!(stream, "{}: {}\n", progname, self.description());
```

Comment:
Yes, that is what Unixy programs are expected to do.  The "//unspecified but not entirely unlike:" comment is meant to give wiggle room to match the expectations of other environments.